src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
I think the memory barrier in pthread_mutex_lock does not avoid the compiler to optimize code by reading variable directly from register (this assembly code is already changed), it is not a memory coherence issue. 
My understanding is, the only way to avoid compiler to do such optimization is to have dependent code of that variable, or have function pointer because compiler cannot understand whether that function pointer will change the variable. However, none of these conditions is clear in the code of CH4 progress engine (e.g., `MPIDI_CH4R_PROGRESS_WHILE`). To be safe, I think it is better to have `volatile` for these variables for now.